utils/ort/src/main/kotlin/storage/S3FileStorage.kt

```diff
@@ -0,0 +1,175 @@
+     */
+    private val accessKeyId: String? = null,
+    /**
+     * The AWS secret for the access key.
+     */
+    private val secretAccessKey: String? = null,
+    /**
+     * Non-public endpoint for testing.
+     */
+    private val customEndpoint: String? = null
+) : FileStorage {
+    private companion object : Logging
+
+    private val scope = CoroutineScope(Dispatchers.Default)
+
+    val s3 by lazy {
+        val provider = if (accessKeyId != null && secretAccessKey != null) {
+            CachedCredentialsProvider(StaticCredentialsProvider(Credentials(accessKeyId, secretAccessKey)))
+        } else {
+            null
+        }
+
+        val client = if (awsRegion != null && provider != null) {
+            S3Client {
+                region = awsRegion
+                credentialsProvider = provider
+                endpointUrl = if (customEndpoint != null) Url.parse(customEndpoint) else null
+            }
+        } else {
+            runBlocking {
+                if (awsRegion != null) {
+                    S3Client.fromEnvironment {
+                        region = awsRegion
+                    }
+                } else {
+                    S3Client.fromEnvironment()
+                }
+            }
+        }
+
+        client
+    }
+
+    override fun exists(path: String): Boolean {
+        val request = HeadObjectRequest {
+            key = path
+            bucket = bucketName
+        }
+
+        var fileExists = true
+
+        runBlocking {
+            try {
+                s3.headObject(request)
+            } catch (e: NotFound) {
+                fileExists = false
+            }
+        }
+
+        return fileExists
+    }
+
+    override fun read(path: String): InputStream {
+        val request = GetObjectRequest {
+            key = path
+            bucket = bucketName
+        }
+
+        val job = scope.async {
+            try {
+                s3.getObject(request) { resp ->
+                    ByteArrayInputStream(resp.body?.toByteArray())
+                }
+            } catch (e: NoSuchKey) {
+                throw NoSuchFileException(File(path))
+            }
+        }
+
+        val stream: InputStream
+
+        runBlocking {
+            stream = job.await()
+        }
+
+        return stream
+    }
+
+    override fun write(path: String, inputStream: InputStream) {
+        val job = scope.launch {
+            inputStream.use {
+                 PutObjectRequest {
+                    key = path
+                    bucket = bucketName
+                    body = ByteStream.fromBytes(it.readBytes())
+                }
+            }.let {
+                try {
+                    s3.putObject(it)
+                } catch (e: S3Exception) {
+                    logger.warn("Can not write $path to S3 bucket $bucketName")
+                }
+            }
+        }
+
+        // Local server doesn't play nice with auto scheduled coroutines. Run blocking instead
```

Comment:
Please properly end (even comment) sentences with a dot, and generally use proper grammar like in a written user's manual, not informal language.

Reply:
Addressed